### PR TITLE
Native login

### PR DESCRIPTION
**Pull request by brandonpage**
### This PR includes:
- suspended `login(username: String, password: String)` function
- `getFallbackWebAuthenticationIntent` function
- `shouldShowBackButton` property/function
- Functions to enable Biometric Authentication via Native Login.
- Added `nativeLogin` property to UserAccount.
- Fix for truly unauthenticated rest requests (necessary for Bio Auth with Native Login).

### Usage: 
SalesforceSDKManager's `initNative` now takes an additional optional native login activity class.

App calls `useNativeLogin` in it's main activity :
```
SalesforceSDKManager.getInstance().useNativeLogin(clientId, redirectUri, loginUrl);
```

In their own UI the App calls:
```
val result = SalesforceSDKManager.getInstance().nativeLoginManager.login("username", "password")
```

### CODE REVIEW

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt:91]: This little trick makes the code super readable.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt:168]: Adding the new param with a default value makes this a non-breaking change in Kotlin, but not Java unfortunately. The new constructor covers that case.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/nativeLoginInterface/NativeLoginManager.kt
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/nativeLoginInterface/NativeLoginManager.kt:51]: Just a reminder:  Unlike `let` in Swift, `val` in Kotlin is **not immutable**.  `val` is simply _readonly_.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/nativeLoginInterface/NativeLoginManager.kt
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/nativeLoginInterface/NativeLoginManager.kt:40]: This class is missing `cancelAuthentication` and `biometricAuthenticationSuccess` compared to it's iOS counterpart because they just aren't necessary on Android.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt:559]: As I mentioned above since we aliased "com.salesforce.androidsdk.auth.**_nativeLoginInterface_**.NativeLoginManager" to `NativeLoginManagerInterface` we can simply use that instead of having to write the _entire path_ for it OR the implementation class of the same name.  And the best part is the consumer still sees the return type as the interface `NativeLoginManager`.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/nativeLoginInterface/NativeLoginManager.kt
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/nativeLoginInterface/NativeLoginManager.kt:73]: I really wanted to handle this for the customer but without an activity (or even if they give me a reference to their activity) it just isn't possible.  Good news is the code the customer will have to implement pretty simple.  Here is an example of the new (`startActivityForResult` is deprecated) way to do it:

```
private val fallbackResult = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
    if (result.resultCode == Activity.RESULT_OK) {
        finish()
    }
}
```

And in their button they simply call `fallbackResult.launch(nativeLoginManager.getFallbackWebAuthenticationIntent())`

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/OAuth2.java
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/OAuth2.java:397]: This feels a bit dirty if you consider that we probably just shouldn't add the interceptor that inserts this authorization header to unauthenticated requests, but attempting that gets pretty complicated.  The client builder is associated with the current account, not the rest client being used.  So even if we are using an unauthenticated rest client, the rest client of the current user is till attached.  

The only scenario in which this matter is Biometric Auth with Native Login.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountBuilder.java
Comment (wmathurin) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountBuilder.java:349]: Wouldn't it be better to call it setNativeLogin??

Comment (brandonpage) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountBuilder.java:349]: I don't have a problem renaming it, but since it's a builder class all of the setters are named this way.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt:140]: I was just noticing this interface and internal implementation looks just like NativeLoginManager where we have the interface an an explicitly `internal` implementation.  `InternalBiometricAuthenticationManager`?

Of course, there's lot of ways to name interface/protocols and implementations 🤯

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt:205]: Both of these could drop their braces, return and `=` a single expression, for optional Kotlin style points.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt:242]: Wouldn't this be safer as an optional rather than a `lateinit`? Since this is public to apps, they could get themselves into a crash if they don't remember to check if this is initialized.  `lateinit` is neat to have, but I've seen enough random crashes leak through that I almost never use it even privately anymore.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt:551]: Simply typo: This should be possessive `its` instead of a contraction for `it is`.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:117]: Could this use a Kotlin string template? This should be equivalent, but perhaps it'd be nice just for consistency.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:189]: I believe we have to eliminate this Context reference in `SalesforceSDKManager` someday.  Should we have native login manager receive it another way now to make that easier?  Perhaps we could have the app provide it in the initialize native login method?  I might be thinking to far ahead on this one.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:213]: `apps` should probably be the possessive `app's`.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:195]: I'm sure this is food for though much later, but it seems unwieldy to have to use `OAuthWebviewHelper` in this context.  I see how it works in the context of the codebase it is today, so no action required.  I have a work item in the backlog where it could be neat to see how we could separate `OAuthWebviewHelper` into smaller units of decoupled logic and then this could use just the one it needs without the two skipped implementations.

Just something to ponder for the future.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:215]: Should `hand` be `hang`?

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt
Comment (JohnsonEricAtSalesforce) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt:242]: Re-reading my own notes thus far, they're all optional except this one.  I do feel we'd save some apps an otherwise obvious crash or two by making this an optional.

**Review by JohnsonEricAtSalesforce — CHANGES_REQUESTED** [review]
Looking great here, overall.  Everything else I noted is optional review, but there's just the one follow-up regarding the `lateinit` on our public Native Login Manager property we should review before merging.

https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/2517#discussion_r1520475464

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt
Comment (brandonpage) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt:91]: You are right on the package name:
```
Names of packages are always lowercase and do not use underscores (org.example.project). Using multi-word names is generally discouraged, but if you do need to use multiple words, you can either just concatenate them together or use the camel case (org.example.myProject).
```
`.native.login.interfaces` is a great suggestion, I see you did that with local notifications.  I'll follow that pattern here too.

Personally I hate class names with `Default` (when there is only one impl) or `Internal` in them because they are just generic junk added on to make the name unique. Might as well by `NativeLoginManager1` for all the value they add lol.  

The reason I like this naming convention is because it is the clearest naming for us **and** the consumer simultaneously.  `auth.native.login.interface.NativeLoginManager` is an interface but to the consumer it is also the implementation.  When we return an instance of that interface to them they get a concrete implementation.  But internally it's just an interface to us, so being able to refer to it as `NativeLoginManagerInterface` internally and have `auth.NativeLoginManager` keep it's name is the best of both worlds imo.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt
Comment (brandonpage) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt:242]: Perhaps.  That seems like a setup issue that customers wouldn't actually see in the wild (similar to a comment I made on your PR earlier).  I feel like customers are either going to use the feature and call `useNativeLogin` in the app's `onCreate` or they aren't going to touch `nativeLoginManager`.  Dealing with optionals is pretty trivial in Kotlin but a bit more annoying in Java.  🤷 

Thoughts on this @wmathurin ?

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt
Comment (brandonpage) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:117]: 🤷 Could be, but I that `loginUrl + OAUTH_AUTH_PATH` is more readable than `"$loginUrl$OAUTH_AUTH_PATH"`.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt
Comment (brandonpage) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:195]: Yeah, I agree.  `OAuthWebviewHelper` is designed around the webview lifecycle and the hooks into it are engrained all over.  I knew it would be too much work for this story/PR to try to separate it into smaller pieces but it would be really cool if we could do that in the future.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountBuilder.java
Comment (wmathurin) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/accounts/UserAccountBuilder.java:349]: That's fine keeping it the same. May have the param name change to isNativeLogin??

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt
Comment (wmathurin) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.kt:236]: Optionals are annoying but compile time failure is still preferable to a run time one (even if unlikely).

**Review by JohnsonEricAtSalesforce — APPROVED** [review]
Thanks for the updates!

**Review by wmathurin — APPROVED** [review]

**Pull request merged**
